File: python/valid_parenthesis.py
class Solution:
    def isValid(self, s: str) -> bool:
        map = {
            ')': '(',
            '}': '{',
            ']': '['
        }

        stack = []

        for c in s:
            if c in map:
                print(stack[-1])
                if len(stack) == 0 or stack[-1] != map[c]:
                    return False
                stack.pop()
            else:
                stack.append(c)

        return len(stack) == 0


s = Solution()
print(s.isValid("([)]{}"))

File: c-sharp/ValidParenthesis.cs
Console.WriteLine(new Solution().IsValid("([)]{}"));

public class Solution
{
    public bool IsValid(string s)
    {
        Stack<char> stack = new();
        Dictionary<char, char> bpair = new() {
            {')', '('},
            {']', '['},
            {'}', '{'}
        };

        foreach (char c in s)
        {
            if (bpair.TryGetValue(c, out char opening))
            {
                if (stack.Count == 0 || stack.Peek() != opening)
                {
                    return false;
                }
                stack.Pop();
            }
            else
            {
                stack.Push(c);
            }
        }

        return stack.Count == 0;
    }
}